File: bindings/dotnet/src/Wcl/WclDocument.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using Wcl.Core;
using Wcl.Eval;
using Wcl.Wasm;

namespace Wcl
{
    public class WclDocument : IDisposable
    {
        private int _handle;
        private bool _disposed;
        private readonly object _lock = new object();

        private OrderedMap<string, WclValue>? _cachedValues;
        private List<Diagnostic>? _cachedDiagnostics;

        internal WclDocument(int handle)
        {
            _handle = handle;
        }

        public OrderedMap<string, WclValue> Values
        {
            get
            {
                lock (_lock)
                {
                    CheckDisposed();
                    if (_cachedValues == null)
                    {
                        var json = WasmRuntime.Instance.DocumentValues(_handle);
                        using var doc = JsonDocument.Parse(json);
                        _cachedValues = JsonConvert.ToValues(doc.RootElement);
                    }
                    return _cachedValues;
                }
            }
        }

        public List<Diagnostic> Diagnostics
        {
            get
            {
                lock (_lock)
                {
                    CheckDisposed();
                    if (_cachedDiagnostics == null)
                    {
                        var json = WasmRuntime.Instance.DocumentDiagnostics(_handle);
                        using var doc = JsonDocument.Parse(json);
                        _cachedDiagnostics = new List<Diagnostic>();
                        foreach (var el in doc.RootElement.EnumerateArray())
                            _cachedDiagnostics.Add(JsonConvert.ToDiagnostic(el));
                    }
                    return _cachedDiagnostics;
                }
            }
        }

        public bool HasErrors()
        {
            lock (_lock)
            {
                CheckDisposed();
                return WasmRuntime.Instance.DocumentHasErrors(_handle);
            }
        }

        public List<Diagnostic> Errors() => Diagnostics.Where(d => d.IsError).ToList();

        public WclValue Query(string query)
        {
            lock (_lock)
            {
                CheckDisposed();
                var resultJson = WasmRuntime.Instance.DocumentQuery(_handle, query);
                using var doc = JsonDocument.Parse(resultJson);
                if (doc.RootElement.TryGetProperty("error", out var errEl))
                    throw new Exception($"query error: {errEl.GetString()}");
                if (doc.RootElement.TryGetProperty("ok", out var okEl))
                    return JsonConvert.ToWclValue(okEl);
                throw new Exception("unexpected query result format");
            }
        }

        public List<BlockRef> Blocks()
        {
            lock (_lock)
            {
                CheckDisposed();
                var json = WasmRuntime.Instance.DocumentBlocks(_handle);
                using var doc = JsonDocument.Parse(json);
                var result = new List<BlockRef>();
                foreach (var el in doc.RootElement.EnumerateArray())
                    result.Add(JsonConvert.ToBlockRef(el));
                return result;
            }
        }

        public List<BlockRef> BlocksOfType(string kind)
        {
            lock (_lock)
            {
                CheckDisposed();
                var json = WasmRuntime.Instance.DocumentBlocksOfType(_handle, kind);
                using var doc = JsonDocument.Parse(json);
                var result = new List<BlockRef>();
                foreach (var el in doc.RootElement.EnumerateArray())
                    result.Add(JsonConvert.ToBlockRef(el));
                return result;
            }
        }

        private void CheckDisposed()
        {
            if (_disposed)
                throw new ObjectDisposedException(nameof(WclDocument));
        }

        public void Dispose()
        {
            lock (_lock)
            {
                if (_disposed) return;
                _disposed = true;

                if (_handle != 0)
                {
                    WasmRuntime.Instance.DocumentFree(_handle);
                    _handle = 0;
                }
            }
            GC.SuppressFinalize(this);
        }

        ~WclDocument()
        {
            Dispose();
        }
    }
}

File: bindings/jvm/src/main/java/io/github/wiltaylor/wcl/WclDocument.java
package io.github.wiltaylor.wcl;

import io.github.wiltaylor.wcl.core.Diagnostic;
import io.github.wiltaylor.wcl.eval.BlockRef;
import io.github.wiltaylor.wcl.eval.WclValue;
import io.github.wiltaylor.wcl.wasm.JsonConvert;
import io.github.wiltaylor.wcl.wasm.WasmRuntime;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.stream.Collectors;

public final class WclDocument implements AutoCloseable {

    private int handle;
    private boolean closed;
    private final Object lock = new Object();

    private LinkedHashMap<String, WclValue> cachedValues;
    private List<Diagnostic> cachedDiagnostics;

    WclDocument(int handle) {
        this.handle = handle;
    }

    public LinkedHashMap<String, WclValue> getValues() {
        synchronized (lock) {
            checkClosed();
            if (cachedValues == null) {
                try {
                    var json = WasmRuntime.getInstance().documentValues(handle);
                    var node = JsonConvert.MAPPER.readTree(json);
                    cachedValues = JsonConvert.toValues(node);
                } catch (Exception e) {
                    throw new WclException("failed to read document values", e);
                }
            }
            return cachedValues;
        }
    }

    public List<Diagnostic> getDiagnostics() {
        synchronized (lock) {
            checkClosed();
            if (cachedDiagnostics == null) {
                try {
                    var json = WasmRuntime.getInstance().documentDiagnostics(handle);
                    var node = JsonConvert.MAPPER.readTree(json);
                    cachedDiagnostics = new ArrayList<>();
                    for (var el : node) {
                        cachedDiagnostics.add(JsonConvert.toDiagnostic(el));
                    }
                } catch (Exception e) {
                    throw new WclException("failed to read diagnostics", e);
                }
            }
            return cachedDiagnostics;
        }
    }

    public boolean hasErrors() {
        synchronized (lock) {
            checkClosed();
            return WasmRuntime.getInstance().documentHasErrors(handle);
        }
    }

    public List<Diagnostic> getErrors() {
        return getDiagnostics().stream().filter(Diagnostic::isError).collect(Collectors.toList());
    }

    public WclValue query(String query) {
        synchronized (lock) {
            checkClosed();
            try {
                var resultJson = WasmRuntime.getInstance().documentQuery(handle, query);
                var node = JsonConvert.MAPPER.readTree(resultJson);
                if (node.has("error")) {
                    throw new WclException("query error: " + node.get("error").textValue());
                }
                if (node.has("ok")) {
                    return JsonConvert.toWclValue(node.get("ok"));
                }
                throw new WclException("unexpected query result format");
            } catch (WclException e) {
                throw e;
            } catch (Exception e) {
                throw new WclException("query failed", e);
            }
        }
    }

    public List<BlockRef> getBlocks() {
        synchronized (lock) {
            checkClosed();
            try {
                var json = WasmRuntime.getInstance().documentBlocks(handle);
                var node = JsonConvert.MAPPER.readTree(json);
                var result = new ArrayList<BlockRef>();
                for (var el : node) result.add(JsonConvert.toBlockRef(el));
                return result;
            } catch (Exception e) {
                throw new WclException("failed to read blocks", e);
            }
        }
    }

    public List<BlockRef> getBlocksOfType(String kind) {
        synchronized (lock) {
            checkClosed();
            try {
                var json = WasmRuntime.getInstance().documentBlocksOfType(handle, kind);
                var node = JsonConvert.MAPPER.readTree(json);
                var result = new ArrayList<BlockRef>();
                for (var el : node) result.add(JsonConvert.toBlockRef(el));
                return result;
            } catch (Exception e) {
                throw new WclException("failed to read blocks of type: " + kind, e);
            }
        }
    }

    private void checkClosed() {
        if (closed) throw new IllegalStateException("WclDocument is closed");
    }

    @Override
    public void close() {
        synchronized (lock) {
            if (closed) return;
            closed = true;
            if (handle != 0) {
                WasmRuntime.getInstance().documentFree(handle);
                handle = 0;
            }
        }
    }

}
